pyvrp/cpp/bindings.cpp

```diff
@@ -27,20 +28,55 @@ using pyvrp::Solution;
 using pyvrp::SubPopulation;
 using TWS = pyvrp::TimeWindowSegment;
 
-template <typename... Args>
-TWS merge(Matrix<pyvrp::Value> const &mat, Args... args)
+namespace pybind11
 {
-    Matrix<pyvrp::Duration> durMat(mat.numRows(), mat.numCols());
-
-    // Copy the Matrix<pyvrp::Value> over to Matrix<Duration>. That's not
-    // efficient, but since this class is internal to PyVRP that does not matter
-    // much. We only expose it to Python for testing.
-    for (size_t row = 0; row != durMat.numRows(); ++row)
-        for (size_t col = 0; col != durMat.numCols(); ++col)
-            durMat(row, col) = mat(row, col);
-
-    return TWS::merge(durMat, args...);
-}
+namespace detail
+{
+template <typename T> struct type_caster<Matrix<T>>
+{
+    static_assert(sizeof(T) == sizeof(pyvrp::Value));
+
+public:
+#ifdef PYVRP_DOUBLE_PRECISION
+    PYBIND11_TYPE_CASTER(Matrix<T>, _("numpy.ndarray[float]"));
+#else
+    PYBIND11_TYPE_CASTER(Matrix<T>, _("numpy.ndarray[int]"));
+#endif
+
```

Comment:
Question: this custom type casting is very new to me. What's the reasoning behind using this over how the bindings were defined previously?

Reply:
Previously we had this `Matrix` thing in Python as well, which was a bit inconvenient to work with because nothing expects that argument. Numpy is much more convenient to work with.

Type casters convert a Python type to a C++ type, and vice versa. That's different from bindings, which merely exposes a C++ type to Python. Here we actually convert two types to/from each other. Now C++ can keep using our `Matrix` class, and pybind will convert numpy arrays to a `Matrix`when needed, or convert a `Matrix` back into a numpy array. To the outside (Python) world, it is as if `Matrix` does not exist: they're all numpy arrays.

Reply:
(I am like or two days of reading ahead of you on this, so I definitely don't know this perfectly either. But it seems to work and that's good enough for me for now!)